MoreLinq/UnFold.cs

```diff
@@ -0,0 +1,87 @@
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Returns a sequence generated by applying a state to the generator function, 
+        /// and from its result, determines if the sequence should have a next element and its value,
+        /// and the next state in the recursive call.
+        /// </summary>
+        /// <typeparam name="TState">Type of state elements.</typeparam>
+        /// <typeparam name="T">Type of the elements generated by the generator function.</typeparam>
+        /// <typeparam name="TResult">The type of the elements of the result sequence.</typeparam>
+        /// <param name="state">The initial state.</param>
+        /// <param name="generator">
+        /// Function that takes a state and produces an object containing all 
+        /// necessary data for the other functions (predicate, stateSelector, resultSelector).
+        /// </param>
+        /// <param name="predicate">
+        /// Function to determine if the unfolding should continue based on the current input.
+        /// </param>
+        /// <param name="stateSelector">
+        /// Function to select the state from the current input.
+        /// </param>
+        /// <param name="resultSelector">
+        /// Function to select the result from the current input.
+        /// </param>
+        /// <returns>A sequence containing the results generated by the resultSelector function.</returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+
+        public static IEnumerable<TResult> Unfold<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            if (generator == null) throw new ArgumentNullException(nameof(generator));
+            if (predicate == null) throw new ArgumentNullException(nameof(predicate));
+            if (stateSelector == null) throw new ArgumentNullException(nameof(stateSelector));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            return UnfoldImpl(state, generator, predicate, stateSelector, resultSelector);
+        }
+
+
+        private static IEnumerable<TResult> UnfoldImpl<TState, T, TResult>(
+            TState state,
+            Func<TState, T> generator,
+            Func<T, bool> predicate,
+            Func<T, TState> stateSelector,
+            Func<T, TResult> resultSelector)
+        {
+            while (true)
+            {
+                var step = generator(state);
+
+                if (!predicate(step))
+                    yield break;
+
+                yield return resultSelector(step);
+                state = stateSelector(step);
```

Comment:
@atifaziz I'm not sure further discussion is helpful at this point. It's pretty clear that its only me that is confused, and a sample size of 1 is not indicative of a problem 😉 .
If I had written this function, however, I would have probably written it more like this:

```csharp
private static IEnumerable<TResult> UnfoldImpl<TState, T, TResult>(
	T seed,
	Func<T, TState> generator,
	Func<TState, bool> predicate,
	Func<TState, T> valueSelector,
	Func<TState, TResult> resultSelector)
{
	T value = seed;
	while (true)
	{
		var state = generator(value);

		if (!predicate(state))
			yield break;

		yield return resultSelector(state);
		value = valueSelector(state);
	}
}
```